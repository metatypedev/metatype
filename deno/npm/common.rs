// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.

/// Gets the corresponding @types package for the provided package name.
pub fn types_package_name(package_name: &str) -> String {
    debug_assert!(!package_name.starts_with("@types/"));
    // Scoped packages will get two underscores for each slash
    // https://github.com/DefinitelyTyped/DefinitelyTyped/tree/15f1ece08f7b498f4b9a2147c2a46e94416ca777#what-about-scoped-packages
    format!("@types/{}", package_name.replace('/', "__"))
}

#[cfg(test)]
mod test {
    use super::types_package_name;

    #[test]
    fn test_types_package_name() {
        assert_eq!(types_package_name("name"), "@types/name");
        assert_eq!(
            types_package_name("@scoped/package"),
            "@types/@scoped__package"
        );
    }
}
