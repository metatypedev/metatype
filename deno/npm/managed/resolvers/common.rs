// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.

use std::collections::HashMap;
use std::io::ErrorKind;
use std::path::Path;
use std::path::PathBuf;
use std::sync::Arc;
use std::sync::Mutex;

use async_trait::async_trait;
use deno_ast::ModuleSpecifier;
use deno_core::error::AnyError;
use deno_core::futures;
use deno_core::unsync::spawn;
use deno_core::url::Url;
use deno_npm::NpmPackageCacheFolderId;
use deno_npm::NpmPackageId;
use deno_npm::NpmResolutionPackage;
use deno_runtime::deno_fs::FileSystem;
use deno_runtime::deno_node::NodePermissions;
use deno_runtime::deno_node::NodeResolutionMode;

use super::super::cache::NpmCache;

/// Part of the resolution that interacts with the file system.
#[async_trait]
pub trait NpmPackageFsResolver: Send + Sync {
    /// Specifier for the root directory.
    fn root_dir_url(&self) -> &Url;

    /// The local node_modules folder if it is applicable to the implementation.
    fn node_modules_path(&self) -> Option<PathBuf>;

    fn package_folder(&self, package_id: &NpmPackageId) -> Result<PathBuf, AnyError>;

    fn resolve_package_folder_from_package(
        &self,
        name: &str,
        referrer: &ModuleSpecifier,
        mode: NodeResolutionMode,
    ) -> Result<PathBuf, AnyError>;

    fn resolve_package_folder_from_specifier(
        &self,
        specifier: &ModuleSpecifier,
    ) -> Result<Option<PathBuf>, AnyError>;

    fn resolve_package_cache_folder_id_from_specifier(
        &self,
        specifier: &ModuleSpecifier,
    ) -> Result<Option<NpmPackageCacheFolderId>, AnyError>;

    async fn cache_packages(&self) -> Result<(), AnyError>;

    fn ensure_read_permission(
        &self,
        permissions: &dyn NodePermissions,
        path: &Path,
    ) -> Result<(), AnyError>;
}

#[derive(Debug)]
pub struct RegistryReadPermissionChecker {
    fs: Arc<dyn FileSystem>,
    cache: Mutex<HashMap<PathBuf, PathBuf>>,
    registry_path: PathBuf,
}

impl RegistryReadPermissionChecker {
    pub fn new(fs: Arc<dyn FileSystem>, registry_path: PathBuf) -> Self {
        Self {
            fs,
            registry_path,
            cache: Default::default(),
        }
    }

    pub fn ensure_registry_read_permission(
        &self,
        permissions: &dyn NodePermissions,
        path: &Path,
    ) -> Result<(), AnyError> {
        // allow reading if it's in the node_modules
        let is_path_in_node_modules = path.starts_with(&self.registry_path)
            && path
                .components()
                .all(|c| !matches!(c, std::path::Component::ParentDir));

        if is_path_in_node_modules {
            let mut cache = self.cache.lock().unwrap();
            let registry_path_canon = match cache.get(&self.registry_path) {
                Some(canon) => canon.clone(),
                None => {
                    let canon = self.fs.realpath_sync(&self.registry_path)?;
                    cache.insert(self.registry_path.to_path_buf(), canon.clone());
                    canon
                }
            };

            let path_canon = match cache.get(path) {
                Some(canon) => canon.clone(),
                None => {
                    let canon = self.fs.realpath_sync(path);
                    if let Err(e) = &canon {
                        if e.kind() == ErrorKind::NotFound {
                            return Ok(());
                        }
                    }

                    let canon = canon?;
                    cache.insert(path.to_path_buf(), canon.clone());
                    canon
                }
            };

            if path_canon.starts_with(registry_path_canon) {
                return Ok(());
            }
        }

        permissions.check_read(path)
    }
}

/// Caches all the packages in parallel.
pub async fn cache_packages(
    packages: Vec<NpmResolutionPackage>,
    cache: &Arc<NpmCache>,
    registry_url: &Url,
) -> Result<(), AnyError> {
    let mut handles = Vec::with_capacity(packages.len());
    for package in packages {
        let cache = cache.clone();
        let registry_url = registry_url.clone();
        let handle = spawn(async move {
            cache
                .ensure_package(&package.id.nv, &package.dist, &registry_url)
                .await
        });
        handles.push(handle);
    }
    let results = futures::future::join_all(handles).await;
    for result in results {
        // surface the first error
        result??;
    }
    Ok(())
}
