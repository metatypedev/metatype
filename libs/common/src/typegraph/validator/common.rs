// Copyright Metatype OÃœ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

use crate::typegraph::{EitherTypeData, TypeNode, Typegraph, UnionTypeData};

impl Typegraph {
    pub fn check_enum_values(
        &self,
        type_idx: u32,
        enum_values: &[String],
    ) -> Result<(), Vec<String>> {
        let type_node = self.types.get(type_idx as usize).unwrap();
        let mut errors = Vec::new();
        if matches!(type_node, TypeNode::Optional { .. }) {
            errors.push("optional not cannot have enumerated values".to_owned());
        } else {
            for value in enum_values {
                match serde_json::from_str::<serde_json::Value>(value) {
                    Ok(val) => match self.validate_value(type_idx, &val) {
                        Ok(_) => {}
                        Err(err) => errors.push(err.to_string()),
                    },
                    Err(e) => errors.push(format!(
                        "Error while deserializing enum value {value:?}: {e:?}"
                    )),
                }
            }
        }

        if errors.is_empty() {
            Ok(())
        } else {
            Err(errors)
        }
    }

    pub fn collect_nested_variants_into(&self, out: &mut Vec<u32>, variants: &[u32]) {
        for idx in variants {
            let node = self.types.get(*idx as usize).unwrap();
            match node {
                TypeNode::Union {
                    data: UnionTypeData { any_of: variants },
                    ..
                }
                | TypeNode::Either {
                    data: EitherTypeData { one_of: variants },
                    ..
                } => self.collect_nested_variants_into(out, variants),
                _ => out.push(*idx),
            }
        }
    }

    pub fn check_union_variants(&self, variants: &[u32]) -> Result<(), String> {
        let mut object_count = 0;

        for variant_type in variants
            .iter()
            .map(|idx| self.types.get(*idx as usize).unwrap())
        {
            match variant_type {
                TypeNode::Object { .. } => object_count += 1,
                TypeNode::Boolean { .. }
                | TypeNode::Float { .. }
                | TypeNode::Integer { .. }
                | TypeNode::String { .. } => {
                    // scalar
                }
                TypeNode::List { data, .. } => {
                    let item_type = self.types.get(data.items as usize).unwrap();
                    if !item_type.is_scalar() {
                        return Err(format!(
                            "array of '{}' not allowed as union/either variant",
                            item_type.type_name()
                        ));
                    }
                }
                _ => {
                    return Err(format!(
                        "type '{}' not allowed as union/either variant",
                        variant_type.type_name()
                    ));
                }
            }
        }

        if object_count != 0 && object_count != variants.len() {
            return Err("union variants must either be all scalars or all objects".to_owned());
        }

        Ok(())
    }
}
