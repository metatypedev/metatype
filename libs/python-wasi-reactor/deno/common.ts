import { WasiReactorOut } from "../bindings/bindings.ts";

type Tag = "ok" | "err";

interface PythonOutput {
  value: string, // json string
  tag: Tag
}

export function processOutput(out: WasiReactorOut): string {
  if ("Ok" in out) {
    // vm output is ok
    const py: PythonOutput = JSON.parse(out.Ok.res);
    if (py.tag == "err") {
      // python error
      throw Error(py.value);
    }
    return py.value;
  }
  // vm error
  throw Error(out.Err.message);
}
