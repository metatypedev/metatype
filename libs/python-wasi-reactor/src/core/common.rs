// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use pyo3::prelude::*;
use pyo3::types::PyTuple;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub enum Tag {
    #[serde(rename = "ok")]
    Ok,
    #[serde(rename = "err")]
    Err,
}

#[derive(Serialize, Deserialize)]
pub struct RetValue {
    pub value: String,
    pub tag: Tag,
}

pub fn collect_args_from_json(args_str: &str) -> Result<Vec<serde_json::Value>, String> {
    let args = serde_json::from_str::<serde_json::Value>(args_str).map_err(|e| e.to_string())?;
    if args.is_array() {
        let res = match args.as_array() {
            Some(arr) => arr.to_vec(),
            None => vec![],
        };
        Ok(res)
    } else {
        Err("string must be a json array".to_string())
    }
}

pub fn pythonize_args(py: Python, pyargs: Vec<serde_json::Value>) -> &PyTuple {
    let eval_args = pyargs
        .iter()
        .map(|value| pythonize::pythonize(py, value).unwrap());
    PyTuple::new(py, eval_args)
}

pub fn recurse_dot_attr(py: &PyModule, name: String) -> PyResult<&PyAny> {
    let parts = name.split('.');
    let mut curr: &PyAny = py;
    for sub in parts {
        curr = curr.getattr(sub)?;
    }
    Ok(curr)
}
