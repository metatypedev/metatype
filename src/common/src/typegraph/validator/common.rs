// Copyright Metatype OÃœ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

use crate::typegraph::{EitherTypeData, TypeNode, Typegraph, UnionTypeData};

impl Typegraph {
    pub fn check_enum_values(
        &self,
        type_idx: u32,
        enum_values: &[String],
    ) -> Result<(), Vec<String>> {
        let type_node = self.types.get(type_idx as usize).unwrap();
        let mut errors = Vec::new();
        if matches!(type_node, TypeNode::Optional { .. }) {
            errors.push("optional not cannot have enumerated values".to_owned());
        } else {
            for value in enum_values {
                match serde_json::from_str::<serde_json::Value>(value) {
                    Ok(val) => match self.validate_value(type_idx, &val) {
                        Ok(_) => {}
                        Err(err) => errors.push(err.to_string()),
                    },
                    Err(e) => errors.push(format!(
                        "Error while deserializing enum value {value:?}: {e:?}"
                    )),
                }
            }
        }

        if errors.is_empty() {
            Ok(())
        } else {
            Err(errors)
        }
    }

    pub fn collect_nested_variants_into(&self, out: &mut Vec<u32>, variants: &[u32]) {
        for idx in variants {
            let node = self.types.get(*idx as usize).unwrap();
            match node {
                TypeNode::Union {
                    data: UnionTypeData { any_of: variants },
                    ..
                }
                | TypeNode::Either {
                    data: EitherTypeData { one_of: variants },
                    ..
                } => self.collect_nested_variants_into(out, variants),
                _ => out.push(*idx),
            }
        }
    }
}
