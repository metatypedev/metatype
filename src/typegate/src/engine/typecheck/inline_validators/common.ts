// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

export function check(condition: string, message: string) {
  return `if (!(${condition})) throw new Error(${message});`;
}
