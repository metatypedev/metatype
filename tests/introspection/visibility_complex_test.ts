// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { Meta } from "test-utils/mod.ts";
import { enumerateTypes, fullIntrospection } from "./common.ts";

const context = {
  control_basic: "PASS",
  control_union: "PASS",
  control_rec_other: "PASS",
  control_inner_variant: "PASS",
};

Meta.test({
  name: "Visibility check upon introspection on complex typegraph",
  introspection: true,
}, async (t) => {
  const e = await t.engine("introspection/visibility_complex.py");
  await t.should(
    "enumerate types without context on complex typegraph",
    async () => {
      await enumerateTypes()
        .matchOkSnapshot(t)
        .on(e);
    },
  );

  await t.should(
    "enumerate types after context defined on complex typegraph",
    async () => {
      await enumerateTypes()
        .withContext(context)
        .matchOkSnapshot(t)
        .on(e);
    },
  );
});

Meta.test({
  name: "Full introspection without context provided on complex typegraph",
  introspection: true,
}, async (t) => {
  const e = await t.engine("introspection/visibility_complex.py");
  await t.should(
    "have a few fields inivible without proper access",
    async () => {
      await fullIntrospection()
        .matchOkSnapshot(t)
        .on(e);
    },
  );
});
