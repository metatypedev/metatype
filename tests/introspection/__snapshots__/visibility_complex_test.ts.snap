export const snapshot = {};

snapshot[`Visibility check upon introspection on complex typegraph 1`] = `
{
  __schema: {
    types: [
      {
        kind: "SCALAR",
        name: "Int",
      },
      {
        kind: "INPUT_OBJECT",
        name: "InputRec_one_struct_depthOne_structInputRec_one_struct_depthOne_struct_Inp",
      },
      {
        kind: "INPUT_OBJECT",
        name: "InputRec_one_structInputRec_one_struct_Inp",
      },
      {
        kind: "INPUT_OBJECT",
        name: "InputRec_two_structInputRec_two_struct_Inp",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output_field_list_optional_struct",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output_field_optional_List_struct",
      },
      {
        kind: "SCALAR",
        name: "String",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output_denied_struct_f1",
      },
      {
        kind: "OBJECT",
        name: "SelfInner",
      },
      {
        kind: "OBJECT",
        name: "SelfRecursive_f2",
      },
      {
        kind: "SCALAR",
        name: "root_identity_read_fn_output_union_scalar_union",
      },
      {
        kind: "SCALAR",
        name: "OnionInner_fempty",
      },
      {
        kind: "OBJECT",
        name: "_variant_empty_OnionInner",
      },
      {
        kind: "SCALAR",
        name: "EmptyStuff",
      },
      {
        kind: "OBJECT",
        name: "_variant_empty_EmptyStuff",
      },
      {
        kind: "UNION",
        name: "UnionWithRecVariant_of_OnionInner_fempty_EmptyStuff_SelfRecursive_f2",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output_f6",
      },
      {
        kind: "OBJECT",
        name: "Query",
      },
      {
        kind: "OBJECT",
        name: "Mutation",
      },
    ],
  },
}
`;

snapshot[`Visibility check upon introspection on complex typegraph 2`] = `
{
  __schema: {
    types: [
      {
        kind: "SCALAR",
        name: "Int",
      },
      {
        kind: "INPUT_OBJECT",
        name: "InputRec_one_struct_depthOne_structInputRec_one_struct_depthOne_struct_Inp",
      },
      {
        kind: "INPUT_OBJECT",
        name: "InputRec_one_structInputRec_one_struct_Inp",
      },
      {
        kind: "INPUT_OBJECT",
        name: "InputRec_two_structInputRec_two_struct_Inp",
      },
      {
        kind: "SCALAR",
        name: "String",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output_field_list_optional_struct",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output_field_optional_List_struct",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output_denied_struct_f1",
      },
      {
        kind: "OBJECT",
        name: "SelfInner",
      },
      {
        kind: "OBJECT",
        name: "SelfRecursive",
      },
      {
        kind: "SCALAR",
        name: "root_identity_read_fn_output_union_scalar_union",
      },
      {
        kind: "OBJECT",
        name: "OnionInner",
      },
      {
        kind: "SCALAR",
        name: "EmptyStuff",
      },
      {
        kind: "OBJECT",
        name: "_variant_empty_EmptyStuff",
      },
      {
        kind: "UNION",
        name: "UnionWithRecVariant_of_OnionInner_EmptyStuff_SelfRecursive",
      },
      {
        kind: "OBJECT",
        name: "root_identity_read_fn_output",
      },
      {
        kind: "OBJECT",
        name: "Query",
      },
      {
        kind: "OBJECT",
        name: "Mutation",
      },
    ],
  },
}
`;

snapshot[`Full introspection without context provided on complex typegraph 1`] = `
{
  __schema: {
    directives: [],
    mutationType: {
      name: "Mutation",
    },
    queryType: {
      name: "Query",
    },
    subscriptionType: null,
    types: [
      {
        description: "integer type",
        enumValues: null,
        fields: null,
        inputFields: null,
        interfaces: [],
        kind: "SCALAR",
        name: "Int",
        possibleTypes: null,
      },
      {
        inputFields: [
          {
            description: "InputRec_one_struct_depthOne_structInputRec_one_struct_depthOne_struct_Inp input field

Policies:
- inherit",
            name: "another",
            type: {
              kind: "NON_NULL",
              name: null,
              ofType: {
                name: "Int",
                ofType: null,
              },
            },
          },
        ],
        interfaces: [],
        kind: "INPUT_OBJECT",
        name: "InputRec_one_struct_depthOne_structInputRec_one_struct_depthOne_struct_Inp",
      },
      {
        inputFields: [
          {
            description: "InputRec_one_structInputRec_one_struct_Inp input field

Policies:
- inherit",
            name: "depthOne",
            type: {
              kind: "NON_NULL",
              name: null,
              ofType: {
                name: "InputRec_one_struct_depthOne_structInputRec_one_struct_depthOne_struct_Inp",
                ofType: null,
              },
            },
          },
        ],
        interfaces: [],
        kind: "INPUT_OBJECT",
        name: "InputRec_one_structInputRec_one_struct_Inp",
      },
      {
        inputFields: [
          {
            description: "InputRec_two_structInputRec_two_struct_Inp input field

Policies:
- denyAll",
            name: "depth_two_one",
            type: {
              kind: "NON_NULL",
              name: null,
              ofType: {
                name: "Int",
                ofType: null,
              },
            },
          },
          {
            description: "InputRec_two_structInputRec_two_struct_Inp input field

Policies:
- inherit",
            name: "depth_two_two",
            type: {
              kind: "LIST",
              name: null,
              ofType: {
                kind: "LIST",
                name: null,
                ofType: {
                  name: "Int",
                  ofType: null,
                },
              },
            },
          },
        ],
        interfaces: [],
        kind: "INPUT_OBJECT",
        name: "InputRec_two_structInputRec_two_struct_Inp",
      },
      {
        fields: [
          {
            args: [],
            description: "root_identity_read_fn_output_field_list_optional_struct field

Policies:
- inherit",
            isDeprecated: false,
            name: "inner_one",
            type: {
              name: "Int",
              ofType: null,
            },
          },
          {
            args: [],
            description: "root_identity_read_fn_output_field_list_optional_struct field

Policies:
- inherit",
            isDeprecated: false,
            name: "inner_two_two",
            type: {
              name: "Int",
              ofType: null,
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "root_identity_read_fn_output_field_list_optional_struct",
      },
      {
        fields: [
          {
            args: [],
            description: "root_identity_read_fn_output_field_optional_List_struct field

Policies:
- inherit",
            isDeprecated: false,
            name: "inner",
            type: {
              name: "Int",
              ofType: null,
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "root_identity_read_fn_output_field_optional_List_struct",
      },
      {
        description: "string type",
        enumValues: null,
        fields: null,
        inputFields: null,
        interfaces: [],
        kind: "SCALAR",
        name: "String",
        possibleTypes: null,
      },
      {
        fields: [
          {
            args: [],
            description: "root_identity_read_fn_output_denied_struct_f1 field

Policies:
- inherit",
            isDeprecated: false,
            name: "field_second",
            type: {
              name: "String",
              ofType: null,
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "root_identity_read_fn_output_denied_struct_f1",
      },
      {
        fields: [
          {
            args: [],
            description: "SelfInner field

Policies:
- inherit",
            isDeprecated: false,
            name: "self_rec",
            type: {
              name: "SelfRecursive_f2",
              ofType: null,
            },
          },
          {
            args: [],
            description: "SelfInner field

Policies:
- inherit",
            isDeprecated: false,
            name: "other",
            type: {
              name: "Int",
              ofType: null,
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "SelfInner",
      },
      {
        fields: [
          {
            args: [],
            description: "SelfRecursive_f2 field

Policies:
- inherit",
            isDeprecated: false,
            name: "self_rec",
            type: {
              name: "SelfRecursive_f2",
              ofType: null,
            },
          },
          {
            args: [],
            description: "SelfRecursive_f2 field

Policies:
- inherit",
            isDeprecated: false,
            name: "inner_self",
            type: {
              kind: "LIST",
              name: null,
              ofType: {
                name: "SelfInner",
                ofType: null,
              },
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "SelfRecursive_f2",
      },
      {
        description: "union type
Int, String",
        kind: "SCALAR",
        name: "root_identity_read_fn_output_union_scalar_union",
      },
      {
        description: "object scalar type representing an empty object",
        kind: "SCALAR",
        name: "OnionInner_fempty",
      },
      {
        description: "_variant_empty_OnionInner type",
        enumValues: null,
        fields: [
          {
            args: [],
            deprecationReason: "Dummy value due to https://github.com/graphql/graphiql/issues/2308",
            isDeprecated: true,
            name: "_",
            type: {
              name: "_variant_empty_OnionInner",
              ofType: null,
            },
          },
        ],
        inputFields: null,
        interfaces: [],
        kind: "OBJECT",
        name: "_variant_empty_OnionInner",
        possibleTypes: null,
      },
      {
        description: "object scalar type representing an empty object",
        kind: "SCALAR",
        name: "EmptyStuff",
      },
      {
        description: "_variant_empty_EmptyStuff type",
        enumValues: null,
        fields: [
          {
            args: [],
            deprecationReason: "Dummy value due to https://github.com/graphql/graphiql/issues/2308",
            isDeprecated: true,
            name: "_",
            type: {
              name: "_variant_empty_EmptyStuff",
              ofType: null,
            },
          },
        ],
        inputFields: null,
        interfaces: [],
        kind: "OBJECT",
        name: "_variant_empty_EmptyStuff",
        possibleTypes: null,
      },
      {
        enumValues: null,
        fields: null,
        inputFields: null,
        interfaces: [],
        kind: "UNION",
        name: "UnionWithRecVariant_of_OnionInner_fempty_EmptyStuff_SelfRecursive_f2",
        possibleTypes: [
          {
            name: "_variant_empty_OnionInner",
            ofType: null,
          },
          {
            name: "_variant_empty_EmptyStuff",
            ofType: null,
          },
          {
            name: "SelfRecursive_f2",
            ofType: null,
          },
        ],
      },
      {
        fields: [
          {
            args: [],
            description: "root_identity_read_fn_output_f6 field

Policies:
- inherit",
            isDeprecated: false,
            name: "field_list_optional",
            type: {
              kind: "LIST",
              name: null,
              ofType: {
                name: "root_identity_read_fn_output_field_list_optional_struct",
                ofType: null,
              },
            },
          },
          {
            args: [],
            description: "root_identity_read_fn_output_f6 field

Policies:
- inherit",
            isDeprecated: false,
            name: "field_optional_List",
            type: {
              kind: "LIST",
              name: null,
              ofType: {
                kind: "LIST",
                name: null,
                ofType: {
                  name: "root_identity_read_fn_output_field_optional_List_struct",
                  ofType: null,
                },
              },
            },
          },
          {
            args: [],
            description: "root_identity_read_fn_output_f6 field

Policies:
- inherit",
            isDeprecated: false,
            name: "denied",
            type: {
              kind: "LIST",
              name: null,
              ofType: {
                kind: "LIST",
                name: null,
                ofType: {
                  name: "root_identity_read_fn_output_denied_struct_f1",
                  ofType: null,
                },
              },
            },
          },
          {
            args: [],
            description: "root_identity_read_fn_output_f6 field

Policies:
- inherit",
            isDeprecated: false,
            name: "mutually_recursive",
            type: {
              kind: "LIST",
              name: null,
              ofType: {
                kind: "LIST",
                name: null,
                ofType: {
                  name: "SelfRecursive_f2",
                  ofType: null,
                },
              },
            },
          },
          {
            args: [],
            description: "root_identity_read_fn_output_f6 field

Policies:
- inherit",
            isDeprecated: false,
            name: "union_scalar",
            type: {
              kind: "NON_NULL",
              name: null,
              ofType: {
                name: "root_identity_read_fn_output_union_scalar_union",
                ofType: null,
              },
            },
          },
          {
            args: [],
            description: "root_identity_read_fn_output_f6 field

Policies:
- inherit",
            isDeprecated: false,
            name: "union_rec_and_empty_variant",
            type: {
              name: "UnionWithRecVariant_of_OnionInner_fempty_EmptyStuff_SelfRecursive_f2",
              ofType: null,
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "root_identity_read_fn_output_f6",
      },
      {
        fields: [
          {
            args: [
              {
                description: "identity_read argument",
                name: "one",
                type: {
                  kind: "NON_NULL",
                  name: null,
                  ofType: {
                    name: "InputRec_one_structInputRec_one_struct_Inp",
                    ofType: null,
                  },
                },
              },
              {
                description: "identity_read argument",
                name: "two",
                type: {
                  kind: "NON_NULL",
                  name: null,
                  ofType: {
                    name: "InputRec_two_structInputRec_two_struct_Inp",
                    ofType: null,
                  },
                },
              },
            ],
            description: "Query field

Policies:
- passThrough",
            isDeprecated: false,
            name: "identity_read",
            type: {
              name: "root_identity_read_fn_output_f6",
              ofType: null,
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "Query",
      },
      {
        fields: [
          {
            args: [
              {
                description: "identity_update argument",
                name: "one",
                type: {
                  kind: "NON_NULL",
                  name: null,
                  ofType: {
                    name: "InputRec_one_structInputRec_one_struct_Inp",
                    ofType: null,
                  },
                },
              },
              {
                description: "identity_update argument",
                name: "two",
                type: {
                  kind: "NON_NULL",
                  name: null,
                  ofType: {
                    name: "InputRec_two_structInputRec_two_struct_Inp",
                    ofType: null,
                  },
                },
              },
            ],
            description: "Mutation field

Policies:
- passThrough",
            isDeprecated: false,
            name: "identity_update",
            type: {
              kind: "LIST",
              name: null,
              ofType: {
                name: "root_identity_read_fn_output_f6",
                ofType: null,
              },
            },
          },
        ],
        interfaces: [],
        kind: "OBJECT",
        name: "Mutation",
      },
    ],
  },
}
`;
